File: Flink0830Tutorial/src/main/scala/com/atguigu/course/AggregateFunctionExample.scala
package com.atguigu.course

import org.apache.flink.api.common.functions.AggregateFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time

// 需求：计算窗口的温度平均值
// 使用增量聚合的方式计算平均值
// 教程 6-11
object AggregateFunctionExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    val sensorData = env.addSource(new SensorSource)
    val avgTempPerWindow = sensorData
      .map(r => (r.id, r.temperature))
      .keyBy(_._1)
      .timeWindow(Time.seconds(15))
      // 注意下面是 aggregate
      .aggregate(new AvgTempFunction)

    avgTempPerWindow.print()

    env.execute()
  }

  class AvgTempFunction extends AggregateFunction[(String, Double), (String, Double, Int), (String, Double)] {
    override def createAccumulator(): (String, Double, Int) = ("", 0.0, 0)

    override def add(value: (String, Double), accumulator: (String, Double, Int)): (String, Double, Int) = {
      // 第一个元素将id保留下来，元组的第二个元素是传感器读数的总数，第三个元素是一共有多少个传感器读数
      (value._1, value._2 + accumulator._2, accumulator._3 + 1)
    }

    override def getResult(accumulator: (String, Double, Int)): (String, Double) = {
      // 第一个元素将 id 保留了下来，第二个元素：平均值 = 温度总和 / 一共有多少个温度读数
      (accumulator._1, accumulator._2 / accumulator._3)
    }

    override def merge(a: (String, Double, Int), b: (String, Double, Int)): (String, Double, Int) = {
      (a._1, a._2 + b._2, a._3 + b._3)
    }
  }
}

File: flink-code-java-0421/src/main/java/com/atguigu/day03/AggregateFunctionExample.java
package com.atguigu.day03;

import com.atguigu.day02.util.SensorReading;
import com.atguigu.day02.util.SensorSource;
import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.time.Time;

public class AggregateFunctionExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<SensorReading> stream = env.addSource(new SensorSource());

        // 分流 => 开窗口 => 聚合
        stream
                .keyBy(r -> r.id)
                .timeWindow(Time.seconds(5))
                .aggregate(new MinAgg())
                .print();

        env.execute();
    }

    public static class MinAgg implements AggregateFunction<SensorReading, Tuple2<String, Double>, Tuple2<String, Double>> {
        @Override
        public Tuple2<String, Double> createAccumulator() {
            return Tuple2.of("", Double.MAX_VALUE);
        }

        @Override
        public Tuple2<String, Double> add(SensorReading r, Tuple2<String, Double> acc) {
            if (r.temperature < acc.f1) {
                return Tuple2.of(r.id, r.temperature);
            } else {
                return acc;
            }
        }

        @Override
        public Tuple2<String, Double> getResult(Tuple2<String, Double> acc) {
            return acc;
        }

        // merge函数在什么时候需要实现？
        // 两个条件：1. 事件时间 2. 会话窗口
        @Override
        public Tuple2<String, Double> merge(Tuple2<String, Double> stringDoubleTuple2, Tuple2<String, Double> acc1) {
            return null;
        }
    }
}
